Store JS Values in IDB

arv opened the pull request:
This contains a few changes that were all related.

- Change keys to `string`s (both for `prolly.Map` and `kv.Store`)
- Removes the flatbuffer for prolly/Leaf and uses JSON instead.
- Changes `kv.Value` to `JSONValue | Uint8Array`
- Changes `prolly.Map` value to `JSONValue` (this has wide implications throughout)
- IDB now stores `JSONValue | Uint8Array` directly

arv commented:
Not sure why the commit history sucks here... Will make sure it is clean

arv commented:
I did a quick comparison storing JS vs storing JSON strings in IDB:

**js in IDB:**

```
populate 1024x1000 (clean, indexes: 0) x 9.70 MB/s ±9.5% (5 runs sampled)
populate 1024x1000 (dirty, indexes: 0) x 11.01 MB/s ±4.3% (6 runs sampled)
populate 1024x1000 (clean, indexes: 1) x 8.53 MB/s ±9.3% (5 runs sampled)
populate 1024x1000 (clean, indexes: 2) x 6.50 MB/s ±3.3% (5 runs sampled)
read tx 1024x1000 x 114.89 MB/s ±780.0% (30 runs sampled)
read tx 1024x5000 x 81.79 MB/s ±1098.0% (5 runs sampled)
scan 1024x1000 x 130.21 MB/s ±1040.0% (29 runs sampled)
scan 1024x5000 x 95.93 MB/s ±1227.7% (5 runs sampled)
write single byte x 0.00 MB/s ±9175.0% (172 runs sampled)
roundtrip write/subscribe/get x 214.01 ops/sec ±2917.2% (110 runs sampled)
create index 1024x1000 x 8.78 ops/sec ±6.9% (5 runs sampled)
create index 1024x5000 x 2.70 ops/sec ±6.1% (5 runs sampled)
subscription 10 x 21.64 ops/sec ±33.4% (11 runs sampled)
subscription 100 x 10.74 ops/sec ±44.8% (6 runs sampled)
subscription 1000 x 10.05 ops/sec ±18.7% (6 runs sampled)
subscription setup 10 x 472.68 ops/sec ±147.6% (237 runs sampled)
subscription setup 100 x 291.59 ops/sec ±109.7% (146 runs sampled)
subscription setup 1000 x 203.59 ops/sec ±371.7% (102 runs sampled)
```

**js -> string in IDB:**

```
populate 1024x1000 (clean, indexes: 0) x 9.15 MB/s ±32.5% (5 runs sampled)
populate 1024x1000 (dirty, indexes: 0) x 9.96 MB/s ±3.3% (6 runs sampled)
populate 1024x1000 (clean, indexes: 1) x 7.81 MB/s ±9.9% (5 runs sampled)
populate 1024x1000 (clean, indexes: 2) x 5.66 MB/s ±4.5% (5 runs sampled)
read tx 1024x1000 x 64.25 MB/s ±438.8% (18 runs sampled)
read tx 1024x5000 x 11.86 MB/s ±78.3% (5 runs sampled)
scan 1024x1000 x 73.98 MB/s ±487.9% (19 runs sampled)
scan 1024x5000 x 12.92 MB/s ±80.3% (5 runs sampled)
write single byte x 0.00 MB/s ±8956.3% (178 runs sampled)
roundtrip write/subscribe/get x 201.13 ops/sec ±2879.3% (103 runs sampled)
create index 1024x1000 x 7.36 ops/sec ±1.8% (5 runs sampled)
create index 1024x5000 x 2.04 ops/sec ±1.3% (5 runs sampled)
subscription 10 x 23.91 ops/sec ±81.7% (12 runs sampled)
subscription 100 x 9.72 ops/sec ±13.2% (5 runs sampled)
subscription 1000 x 10.26 ops/sec ±39.6% (6 runs sampled)
subscription setup 10 x 424.47 ops/sec ±126.1% (213 runs sampled)
subscription setup 100 x 302.01 ops/sec ±87.1% (152 runs sampled)
subscription setup 1000 x 219.25 ops/sec ±135.6% (110 runs sampled)
```

arv commented:
The results are not very stable. I think I need to redo the tests with everything shut down...

I also want to test js -> json string -> uint8array

aboodman commented:
When you provide these benchmarks can you please provide the baseline (no
change) too? I can't remember where we are currently at and it depends on
your machine of course too.

On Thu, Sep 2, 2021 at 7:26 AM Erik Arvidsson ***@***.***>
wrote:

> The results are not very stable. I think I need to redo the tests with
> everything shut down...
>
> I also want to test js -> json string -> uint8array
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/rocicorp/replicache/pull/470#issuecomment-911904324>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAATUBDCJSW754QHPMBPXJ3T76XV3ANCNFSM5DJRW36Q>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>

arv commented:
Here is a all the different branches all run in isolation (everything else shut down).


**main:** `put(flatbufferEncode(value))`

```
populate 1024x1000 (clean, indexes: 0) x 8.14 MB/s ±9.8% (5 runs sampled)
populate 1024x1000 (dirty, indexes: 0) x 8.80 MB/s ±6.4% (5 runs sampled)
populate 1024x1000 (clean, indexes: 1) x 5.97 MB/s ±7.7% (5 runs sampled)
populate 1024x1000 (clean, indexes: 2) x 4.51 MB/s ±7.3% (5 runs sampled)
read tx 1024x1000 x 25.83 MB/s ±158.7% (9 runs sampled)
read tx 1024x5000 x 11.17 MB/s ±67.8% (5 runs sampled)
scan 1024x1000 x 29.68 MB/s ±174.2% (10 runs sampled)
scan 1024x5000 x 11.47 MB/s ±71.5% (5 runs sampled)
write single byte x 0.00 MB/s ±9506.7% (179 runs sampled)
roundtrip write/subscribe/get x 209.75 ops/sec ±3071.4% (108 runs sampled)
create index 1024x1000 x 6.71 ops/sec ±11.3% (5 runs sampled)
create index 1024x5000 x 1.88 ops/sec ±3.5% (5 runs sampled)
subscription 10 x 3.27 ops/sec ±6.8% (5 runs sampled)
subscription 100 x 3.26 ops/sec ±3.2% (5 runs sampled)
subscription 1000 x 3.23 ops/sec ±15.5% (5 runs sampled)
subscription setup 10 x 181.67 ops/sec ±88.7% (91 runs sampled)
subscription setup 100 x 86.38 ops/sec ±22.2% (44 runs sampled)
subscription setup 1000 x 64.44 ops/sec ±65.6% (33 runs sampled)
```

**uint8array-in-idb:** `put(utf8.encode(JSON.stringify(value)))`

```
populate 1024x1000 (clean, indexes: 0) x 9.63 MB/s ±11.0% (5 runs sampled)
populate 1024x1000 (dirty, indexes: 0) x 9.69 MB/s ±5.1% (5 runs sampled)
populate 1024x1000 (clean, indexes: 1) x 8.97 MB/s ±10.5% (5 runs sampled)
populate 1024x1000 (clean, indexes: 2) x 6.61 MB/s ±5.5% (5 runs sampled)
read tx 1024x1000 x 37.13 MB/s ±244.1% (11 runs sampled)
read tx 1024x5000 x 11.90 MB/s ±73.5% (5 runs sampled)
scan 1024x1000 x 39.54 MB/s ±253.0% (12 runs sampled)
scan 1024x5000 x 12.16 MB/s ±74.6% (5 runs sampled)
write single byte x 0.00 MB/s ±9446.7% (182 runs sampled)
roundtrip write/subscribe/get x 224.92 ops/sec ±3067.9% (115 runs sampled)
create index 1024x1000 x 7.98 ops/sec ±6.8% (5 runs sampled)
create index 1024x5000 x 2.41 ops/sec ±1.1% (5 runs sampled)
subscription 10 x 21.72 ops/sec ±182.0% (11 runs sampled)
subscription 100 x 10.17 ops/sec ±8.8% (6 runs sampled)
subscription 1000 x 9.83 ops/sec ±4.9% (5 runs sampled)
subscription setup 10 x 385.46 ops/sec ±804.3% (193 runs sampled)
subscription setup 100 x 289.19 ops/sec ±103.2% (145 runs sampled)
subscription setup 1000 x 197.33 ops/sec ±129.8% (99 runs sampled)
```

**strings-in-idb:** `put(JSON.stringify(value))`

```
populate 1024x1000 (clean, indexes: 0) x 9.45 MB/s ±9.7% (5 runs sampled)
populate 1024x1000 (dirty, indexes: 0) x 10.25 MB/s ±3.7% (6 runs sampled)
populate 1024x1000 (clean, indexes: 1) x 7.80 MB/s ±8.9% (5 runs sampled)
populate 1024x1000 (clean, indexes: 2) x 5.72 MB/s ±6.6% (5 runs sampled)
read tx 1024x1000 x 65.98 MB/s ±455.4% (17 runs sampled)
read tx 1024x5000 x 13.24 MB/s ±82.6% (5 runs sampled)
scan 1024x1000 x 73.43 MB/s ±489.5% (19 runs sampled)
scan 1024x5000 x 13.53 MB/s ±82.7% (5 runs sampled)
write single byte x 0.00 MB/s ±8775.0% (177 runs sampled)
roundtrip write/subscribe/get x 209.75 ops/sec ±3662.1% (105 runs sampled)
create index 1024x1000 x 7.71 ops/sec ±13.2% (5 runs sampled)
create index 1024x5000 x 2.13 ops/sec ±6.9% (5 runs sampled)
subscription 10 x 21.70 ops/sec ±79.3% (11 runs sampled)
subscription 100 x 10.09 ops/sec ±18.3% (6 runs sampled)
subscription 1000 x 10.66 ops/sec ±11.5% (6 runs sampled)
subscription setup 10 x 441.29 ops/sec ±942.9% (221 runs sampled)
subscription setup 100 x 287.70 ops/sec ±112.5% (145 runs sampled)
subscription setup 1000 x 195.38 ops/sec ±200.0% (99 runs sampled)
```

**js-values-in-idb:** `put(value)`

```
populate 1024x1000 (clean, indexes: 0) x 9.98 MB/s ±12.8% (5 runs sampled)
populate 1024x1000 (dirty, indexes: 0) x 11.28 MB/s ±3.5% (6 runs sampled)
populate 1024x1000 (clean, indexes: 1) x 8.78 MB/s ±7.5% (5 runs sampled)
populate 1024x1000 (clean, indexes: 2) x 6.73 MB/s ±6.8% (5 runs sampled)
read tx 1024x1000 x 119.09 MB/s ±797.6% (31 runs sampled)
read tx 1024x5000 x 14.71 MB/s ±88.7% (5 runs sampled)
scan 1024x1000 x 119.09 MB/s ±765.9% (33 runs sampled)
scan 1024x5000 x 14.82 MB/s ±90.3% (5 runs sampled)
write single byte x 0.00 MB/s ±8781.3% (180 runs sampled)
roundtrip write/subscribe/get x 216.00 ops/sec ±2941.4% (111 runs sampled)
create index 1024x1000 x 8.88 ops/sec ±2.5% (5 runs sampled)
create index 1024x5000 x 2.75 ops/sec ±5.6% (5 runs sampled)
subscription 10 x 19.41 ops/sec ±58.1% (10 runs sampled)
subscription 100 x 9.97 ops/sec ±9.9% (5 runs sampled)
subscription 1000 x 9.93 ops/sec ±9.4% (5 runs sampled)
subscription setup 10 x 452.64 ops/sec ±128.6% (227 runs sampled)
subscription setup 100 x 298.45 ops/sec ±116.1% (150 runs sampled)
subscription setup 1000 x 211.45 ops/sec ±104.3% (106 runs sampled)
```

arv committed:
Store JS Values in IDB

This contains a few changes that were all related.

- Change keys to `string`s (both for `prolly.Map` and `kv.Store`)
- Removes the flatbuffer for prolly/Leaf and uses JSON instead.
- Changes `kv.Value` to `JSONValue | Uint8Array`
- Changes `prolly.Map` value to `JSONValue` (this has wide implications throughout)
- IDB now stores `JSONValue | Uint8Array` directly

arv committed:
No need to JSON stringify console log

arv commented on src/hash.ts:
```diff
@@ -17,7 +17,7 @@ export class Hash {
    * You have to await the result of [[initHasher]] before calling this method.
    */
-  static of(sum: Uint8Array): Hash {
+  static of(sum: Uint8Array | string): Hash {
```
hash-wasm supports hashing a string directly 👍

aboodman commented:
>
> subscription setup 10 x 181.67 ops/sec ±88.7% (91 runs sampled)
> subscription setup 100 x 86.38 ops/sec ±22.2% (44 runs sampled)
> subscription setup 1000 x 64.44 ops/sec ±65.6% (33 runs sampled)
> ...
> subscription setup 10 x 452.64 ops/sec ±128.6% (227 runs sampled)
> subscription setup 100 x 298.45 ops/sec ±116.1% (150 runs sampled)
> subscription setup 1000 x 211.45 ops/sec ±104.3% (106 runs sampled)


lol.

On Thu, Sep 2, 2021 at 10:57 AM Erik Arvidsson ***@***.***>
wrote:

> Closed #470 <https://github.com/rocicorp/replicache/pull/470>.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/rocicorp/replicache/pull/470#event-5249984901>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAATUBBFDJZ562LVIOS6AQDT77QMVANCNFSM5DJRW36Q>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>

aboodman commented:
It would be really nice to output these benchmarks in some format where we
can more easily dump them in a table or something to see the effects. It's
really hard to parse this way.

On Thu, Sep 2, 2021 at 11:12 AM Aaron Boodman ***@***.***>
wrote:

> subscription setup 10 x 181.67 ops/sec ±88.7% (91 runs sampled)
>> subscription setup 100 x 86.38 ops/sec ±22.2% (44 runs sampled)
>> subscription setup 1000 x 64.44 ops/sec ±65.6% (33 runs sampled)
>> ...
>> subscription setup 10 x 452.64 ops/sec ±128.6% (227 runs sampled)
>> subscription setup 100 x 298.45 ops/sec ±116.1% (150 runs sampled)
>> subscription setup 1000 x 211.45 ops/sec ±104.3% (106 runs sampled)
>
>
> lol.
>
> On Thu, Sep 2, 2021 at 10:57 AM Erik Arvidsson ***@***.***>
> wrote:
>
>> Closed #470 <https://github.com/rocicorp/replicache/pull/470>.
>>
>> —
>> You are receiving this because you commented.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/rocicorp/replicache/pull/470#event-5249984901>, or
>> unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/AAATUBBFDJZ562LVIOS6AQDT77QMVANCNFSM5DJRW36Q>
>> .
>> Triage notifications on the go with GitHub Mobile for iOS
>> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
>> or Android
>> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>>
>>
>

arv reviewed (commented):
> It would be really nice to output these benchmarks in some format where we can more easily dump them in a table or something to see the effects. It's really hard to parse this way.

I agree. I've been thinking about a tool where I can copy paste the results in a few text boxes and it would extract the data and show relative speed

arv commented:
Another annoying thing is that the variance is huge on some of these tests which makes it hard to draw any kind of conclusion

arv committed:
Cleanup

aboodman commented:
Yeah I get the impression that these tests are no longer measuring the most
important thing -- that they are being swamped by something else.

On Thu, Sep 2, 2021 at 11:17 AM Erik Arvidsson ***@***.***>
wrote:

> Another annoying thing is that the variance is huge on some of these tests
> which makes it hard to draw any kind of conclusion
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/rocicorp/replicache/pull/470#issuecomment-912064285>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAATUBA7ZAIYOTFKH5BP7F3T77SXJANCNFSM5DJRW36Q>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>
